libraries/standard/http/src/http_client.c

```diff
@@ -1,42 +1,311 @@
+                             HTTP_USER_AGENT_FIELD_LEN,
+                             HTTP_USER_AGENT_VALUE,
+                             STRLEN_LITERAL( HTTP_USER_AGENT_VALUE ) );
+    }
+
+    if( HTTP_SUCCEEDED( status ) )
+    {
+        pRequestHeaders->pBuffer += toAddLen;
+        /* Write "Host: <Value>". */
+        status = _addHeader( pRequestHeaders,
+                             HTTP_HOST_FIELD,
+                             HTTP_HOST_FIELD_LEN,
+                             pRequestInfo->pHost,
+                             pRequestInfo->hostLen );
+    }
+
+    if( HTTP_SUCCEEDED( status ) )
+    {
+        if( HTTP_REQUEST_KEEP_ALIVE_FLAG & pRequestInfo->flags )
+        {
+            /* Write "Connection: keep-alive" */
+            status = _addHeader( pRequestHeaders,
+                                 HTTP_CONNECTION_FIELD,
+                                 HTTP_CONNECTION_FIELD_LEN,
+                                 HTTP_CONNECTION_KEEP_ALIVE_VALUE,
+                                 HTTP_CONNECTION_KEEP_ALIVE_VALUE_LEN );
+        }
+        else
+        {
+            /* Write "Connection: close" */
+            status = _addHeader( pRequestHeaders,
+                                 HTTP_CONNECTION_FIELD,
+                                 HTTP_CONNECTION_FIELD_LEN,
+                                 HTTP_CONNECTION_CLOSE_VALUE,
+                                 HTTP_CONNECTION_CLOSE_VALUE_LEN );
+        }
+    }
+
+    if( HTTP_SUCCEEDED( status ) )
+    {
+        /* Write "\r\n" line to end the HTTP header. */
+        if( ( HTTP_HEADER_LINE_SEPARATOR_LEN + pRequestHeaders->headersLen )
+            > pRequestHeaders->bufferLen )
+        {
+            status = HTTP_INSUFFICIENT_MEMORY;
+        }
+        else
+        {
+            /* TODO: Add log. */
+        }
+    }
+
+    if( HTTP_SUCCEEDED( status ) )
+    {
+        memcpy( pBufferCur,
+                HTTP_HEADER_LINE_SEPARATOR,
+                HTTP_HEADER_LINE_SEPARATOR_LEN );
+        pRequestHeaders->headersLen += HTTP_HEADER_LINE_SEPARATOR_LEN;
+    }
+
+    return status;
+}
+
+HTTPStatus_t HTTPClient_AddHeader( HTTPRequestHeaders_t * pRequestHeaders,
+                                   const char * pField,
+                                   size_t fieldLen,
+                                   const char * pValue,
+                                   size_t valueLen )
+{
+    HTTPStatus_t status = HTTP_INTERNAL_ERROR;
+
+    /* Check if header field is long enough for length to overflow. */
+    if( fieldLen > ( UINT32_MAX >> 2 ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* Check if header value is long enough for length to overflow. */
+    if( valueLen > ( UINT32_MAX >> 2 ) )
+    {
+        /* TODO: Add log. */
+        status = HTTP_INVALID_PARAMETER;
+    }
+
+    /* "Content-Length" header must not be set by user if
+     * HTTP_REQUEST_DISABLE_CONTENT_LENGTH_FLAG is deactivated. */
```

Comment:
How are we going to check for that flag here? Looks like I made an oversight on how to carry that state to other public function.

Reply:
EXACTLY! I realized that we needed pRequestInfo to check for that flag but that would be quite inconvenient for the user.

Reply:
We could add the flags to the RequestHeaders_t structure.